### PR TITLE
user route

### PR DESCRIPTION
**Pull request by WangBoxue**

### CODE REVIEW

File: routes/user.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+  # retrieve all users
+  get '/users' do
+    User.all.to_json
+  end
+
+  # retrieve top ten users sorted by point
+  get '/users/topten' do
+    User.order(point: :desc).limit(10).to_json
+  end
```
Comment (liuyang1204) [inline routes/user.rb]: What about:

```
GET /users/rank            // default limit=10
GET /users/rank?limit=20
```

File: routes/user.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+  # retrieve one user
+  get '/users/:user_id' do
+      User.find(params[:user_id]).to_json
+  end
```
Comment (liuyang1204) [inline routes/user.rb]: I feel these 2 APIs' URLs are coupled.

What if a user's id is "topten"?

Comment (liuyang1204) [inline routes/user.rb]: One way is to change the rank API to `POST`

File: routes/user.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  # retrieve speeches this user have attended
+  get '/users/:user_id/attended_speeches' do
+    User.find(params[:user_id]).attended_speeches.to_json
+  end
+
+  # retrieve exchange history, including prize information
+  get '/users/:user_id/exchange_history' do
+    User.find(params[:user_id]).exchanges.to_json(include: :prize)
+  end
```
Comment (liuyang1204) [inline routes/user.rb]: REST api uses `resource name`. Here I feel better to use 

GET /users/:user_id/exchanges

File: routes/user.rb (hunk 2 above)
Comment (WangBoxue) [inline routes/user.rb]: user id must be integer

File: routes/user.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    user.save!
+    user.to_json
+  end
+
+  # update the point of a user
+  put '/users/:user_id' do
+    user = User.find(params[:user_id])
+    user.update(point: @body['point'])
+    user.to_json
+  end
```
Comment (liuyang1204) [inline routes/user.rb]: I don't think this API is necessary, it is very dangerous.

The points of users should be changed by other actions, e.x. The admin confirm the end of a speech, and input the actual audience list, and click "confirm". In the `confirm` action, we can change all users' points.

File: routes/user.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+  # user rank sorted by point
+  # retrieve ten users by default
+  # you can use a query parameter 'limit' to retrieve a specific amount of user
+  post '/users/rank' do
+    limit = params[:limit]
+    if limit.nil?
+      limit = 10
+    end
+    User.order(point: :desc).limit(limit).to_json
```
Comment (liuyang1204) [inline routes/user.rb]: you can short it to 

```
limit = params[:limit] || 10
```

File: routes/user.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  get '/users/:user_id/exchanges' do
+    User.find(params[:user_id]).exchanges.to_json(include: :prize)
+  end
+
+  # add a normal user
+  post '/users' do
+    user = User.new(id: @body['id'], role: 'user', point: 0)
+    user.save!
+    user.to_json
+  end
```
Comment (liuyang1204) [inline routes/user.rb]: I can not image when we need this API.

Comment (WangBoxue) [inline routes/user.rb]: When a new user comes to this plugin, don't we need to add this user?

**Pull request merged**